Library/Homebrew/dev-cmd/vendor-gems.rb

```diff
@@ -28,13 +28,31 @@ def vendor_gems_args
   def vendor_gems
     args = vendor_gems_args.parse
 
-    Homebrew.install_bundler!
+    rubies = {}
+    case (major_minor_ruby = Version.new(RUBY_VERSION).major_minor)
+    when "2.6"
+      rubies["2.6"] = RUBY_PATH
+      rubies["3.1"] = ensure_formula_installed!("ruby@3.1", reason: "Ruby 3.1 vendoring").opt_bin/"ruby"
```

Comment:
> Surely these files are actually the same regardless of the Ruby version? I can't see a good reason to duplicate them in multiple parts of the file tree.

For the stuff we commit: yeah.

It is however not safe to symlink the entire directory:

<details>
<summary>Diff</summary>

```diff
 {2.6.0 => 3.1.0}/gems/bootsnap-1.13.0/ext/bootsnap/Makefile                                                                       |   73 +++----
 {2.6.0 => 3.1.0}/gems/bootsnap-1.13.0/ext/bootsnap/bootsnap.bundle                                                                |  Bin 38832 -> 38840 bytes
 {2.6.0 => 3.1.0}/gems/bootsnap-1.13.0/ext/bootsnap/bootsnap.o                                                                     |  Bin 51408 -> 65520 bytes
 {2.6.0 => 3.1.0}/gems/bootsnap-1.13.0/lib/bootsnap/bootsnap.bundle                                                                |  Bin 38832 -> 38840 bytes
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/ext/byebug/Makefile                                                                           |   73 +++----
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/ext/byebug/breakpoint.o                                                                       |  Bin 31568 -> 48808 bytes
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/ext/byebug/byebug.bundle                                                                      |  Bin 88624 -> 87480 bytes
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/ext/byebug/byebug.o                                                                           |  Bin 53624 -> 68200 bytes
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/ext/byebug/context.o                                                                          |  Bin 52624 -> 72440 bytes
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/ext/byebug/locker.o                                                                           |  Bin 5344 -> 4648 bytes
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/ext/byebug/threads.o                                                                          |  Bin 15592 -> 19176 bytes
 {2.6.0 => 3.1.0}/gems/byebug-11.1.3/lib/byebug/byebug.bundle                                                                      |  Bin 88624 -> 87480 bytes
 2.6.0/gems/did_you_mean-1.6.1/.github/workflows/ruby.yml => /dev/null                                                             |   39 ----
 2.6.0/gems/did_you_mean-1.6.1/.gitignore => /dev/null                                                                             |   19 --
 2.6.0/gems/did_you_mean-1.6.1/CHANGELOG.md => /dev/null                                                                           |  547 ------------------------------------------------
 2.6.0/gems/did_you_mean-1.6.1/Gemfile => /dev/null                                                                                |   10 -
 2.6.0/gems/did_you_mean-1.6.1/LICENSE.txt => /dev/null                                                                            |   22 --
 2.6.0/gems/did_you_mean-1.6.1/README.md => /dev/null                                                                              |  165 ---------------
 2.6.0/gems/did_you_mean-1.6.1/Rakefile => /dev/null                                                                               |   73 -------
 2.6.0/gems/did_you_mean-1.6.1/appveyor.yml => /dev/null                                                                           |   25 ---
 2.6.0/gems/did_you_mean-1.6.1/benchmark/jaro_winkler/memory_usage.rb => /dev/null                                                 |   12 --
 2.6.0/gems/did_you_mean-1.6.1/benchmark/jaro_winkler/speed.rb => /dev/null                                                        |   22 --
 2.6.0/gems/did_you_mean-1.6.1/benchmark/levenshtein/memory_usage.rb => /dev/null                                                  |   12 --
 2.6.0/gems/did_you_mean-1.6.1/benchmark/levenshtein/speed.rb => /dev/null                                                         |   22 --
 2.6.0/gems/did_you_mean-1.6.1/benchmark/memory_usage.rb => /dev/null                                                              |   16 --
 2.6.0/gems/did_you_mean-1.6.1/benchmark/require_path_checker.rb => /dev/null                                                      |   47 -----
 2.6.0/gems/did_you_mean-1.6.1/benchmark/speed.yml => /dev/null                                                                    |   13 --
 2.6.0/gems/did_you_mean-1.6.1/did_you_mean.gemspec => /dev/null                                                                   |   27 ---
 2.6.0/gems/did_you_mean-1.6.1/documentation/CHANGELOG.md.erb => /dev/null                                                         |    8 -
 2.6.0/gems/did_you_mean-1.6.1/documentation/changelog_generator.rb => /dev/null                                                   |   34 ---
 2.6.0/gems/did_you_mean-1.6.1/documentation/human_typo_api.md => /dev/null                                                        |   20 --
 2.6.0/gems/did_you_mean-1.6.1/documentation/tree_spell_algorithm.md => /dev/null                                                  |   82 --------
 2.6.0/gems/did_you_mean-1.6.1/documentation/tree_spell_checker_api.md => /dev/null                                                |   24 ---
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/core_ext/name_error.rb => /dev/null                                                |   32 ---
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/experimental.rb => /dev/null                                                       |    2 -
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/formatter.rb => /dev/null                                                          |   44 ----
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/formatters/plain_formatter.rb => /dev/null                                         |    4 -
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/formatters/verbose_formatter.rb => /dev/null                                       |    9 -
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/jaro_winkler.rb => /dev/null                                                       |   87 --------
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/levenshtein.rb => /dev/null                                                        |   57 -----
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checker.rb => /dev/null                                                      |   46 -----
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/key_error_checker.rb => /dev/null                                   |   20 --
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/method_name_checker.rb => /dev/null                                 |   72 -------
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb => /dev/null              |   49 -----
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/name_error_checkers/variable_name_checker.rb => /dev/null           |   85 --------
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/name_error_checkers.rb => /dev/null                                 |   20 --
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/null_checker.rb => /dev/null                                        |    6 -
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/pattern_key_name_checker.rb => /dev/null                            |   20 --
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/spell_checkers/require_path_checker.rb => /dev/null                                |   39 ----
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/tree_spell_checker.rb => /dev/null                                                 |  109 ----------
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/verbose.rb => /dev/null                                                            |    2 -
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean/version.rb => /dev/null                                                            |    3 -
 2.6.0/gems/did_you_mean-1.6.1/lib/did_you_mean.rb => /dev/null                                                                    |  155 --------------
 2.6.0/gems/did_you_mean-1.6.1/test/core_ext/test_name_error_extension.rb => /dev/null                                             |   49 -----
 2.6.0/gems/did_you_mean-1.6.1/test/edit_distance/test_jaro_winkler.rb => /dev/null                                                |   36 ----
 2.6.0/gems/did_you_mean-1.6.1/test/fixtures/book.rb => /dev/null                                                                  |    4 -
 2.6.0/gems/did_you_mean-1.6.1/test/fixtures/mini_dir.yml => /dev/null                                                             |   15 --
 2.6.0/gems/did_you_mean-1.6.1/test/fixtures/rspec_dir.yml => /dev/null                                                            |  112 ----------
 2.6.0/gems/did_you_mean-1.6.1/test/helper.rb => /dev/null                                                                         |   33 ---
 2.6.0/gems/did_you_mean-1.6.1/test/spell_checking/test_class_name_check.rb => /dev/null                                           |   81 --------
 2.6.0/gems/did_you_mean-1.6.1/test/spell_checking/test_key_name_check.rb => /dev/null                                             |   54 -----
 2.6.0/gems/did_you_mean-1.6.1/test/spell_checking/test_method_name_check.rb => /dev/null                                          |  147 -------------
 2.6.0/gems/did_you_mean-1.6.1/test/spell_checking/test_pattern_key_name_check.rb => /dev/null                                     |   20 --
 2.6.0/gems/did_you_mean-1.6.1/test/spell_checking/test_require_path_check.rb => /dev/null                                         |   32 ---
 2.6.0/gems/did_you_mean-1.6.1/test/spell_checking/test_uncorrectable_name_check.rb => /dev/null                                   |   15 --
 2.6.0/gems/did_you_mean-1.6.1/test/spell_checking/test_variable_name_check.rb => /dev/null                                        |  140 -------------
 2.6.0/gems/did_you_mean-1.6.1/test/test_ractor_compatibility.rb => /dev/null                                                      |  102 ---------
 2.6.0/gems/did_you_mean-1.6.1/test/test_spell_checker.rb => /dev/null                                                             |   78 -------
 2.6.0/gems/did_you_mean-1.6.1/test/test_tree_spell_checker.rb => /dev/null                                                        |  170 ---------------
 2.6.0/gems/did_you_mean-1.6.1/test/tree_spell/change_word.rb => /dev/null                                                         |   61 ------
 2.6.0/gems/did_you_mean-1.6.1/test/tree_spell/human_typo.rb => /dev/null                                                          |   69 -------
 2.6.0/gems/did_you_mean-1.6.1/test/tree_spell/test_change_word.rb => /dev/null                                                    |   38 ----
 2.6.0/gems/did_you_mean-1.6.1/test/tree_spell/test_explore.rb => /dev/null                                                        |  128 ------------
 2.6.0/gems/did_you_mean-1.6.1/test/tree_spell/test_human_typo.rb => /dev/null                                                     |   24 ---
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/ext/fast_xs/Makefile                                                                          |   73 +++----
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/ext/fast_xs/fast_xs.bundle                                                                    |  Bin 35048 -> 35000 bytes
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/ext/fast_xs/fast_xs.o                                                                         |  Bin 15416 -> 16824 bytes
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/ext/hpricot_scan/Makefile                                                                     |   73 +++----
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/ext/hpricot_scan/hpricot_css.o                                                                |  Bin 86568 -> 95048 bytes
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/ext/hpricot_scan/hpricot_scan.bundle                                                          |  Bin 224192 -> 254576 bytes
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/ext/hpricot_scan/hpricot_scan.o                                                               |  Bin 640744 -> 731680 bytes
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/lib/fast_xs.bundle                                                                            |  Bin 35048 -> 35000 bytes
 {2.6.0 => 3.1.0}/gems/hpricot-0.8.6/lib/hpricot_scan.bundle                                                                       |  Bin 224192 -> 254576 bytes
 {2.6.0 => 3.1.0}/gems/json-2.6.2/ext/json/Makefile                                                                                |   73 +++----
 {2.6.0 => 3.1.0}/gems/json-2.6.2/ext/json/ext/generator/Makefile                                                                  |   73 +++----
 {2.6.0 => 3.1.0}/gems/json-2.6.2/ext/json/ext/generator/generator.bundle                                                          |  Bin 63280 -> 62848 bytes
 {2.6.0 => 3.1.0}/gems/json-2.6.2/ext/json/ext/generator/generator.o                                                               |  Bin 118616 -> 181872 bytes
 {2.6.0 => 3.1.0}/gems/json-2.6.2/ext/json/ext/parser/Makefile                                                                     |   73 +++----
 {2.6.0 => 3.1.0}/gems/json-2.6.2/ext/json/ext/parser/parser.bundle                                                                |  Bin 40496 -> 56768 bytes
 {2.6.0 => 3.1.0}/gems/json-2.6.2/ext/json/ext/parser/parser.o                                                                     |  Bin 77264 -> 86296 bytes
 {2.6.0 => 3.1.0}/gems/json-2.6.2/lib/json/ext/generator.bundle                                                                    |  Bin 63280 -> 62848 bytes
 {2.6.0 => 3.1.0}/gems/json-2.6.2/lib/json/ext/parser.bundle                                                                       |  Bin 40496 -> 56768 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/Makefile                                                                          |   75 +++----
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/buffer.o                                                                          |  Bin 49408 -> 57392 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/buffer_class.o                                                                    |  Bin 60664 -> 65224 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/extension_value_class.o                                                           |  Bin 4240 -> 3864 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/factory_class.o                                                                   |  Bin 34576 -> 35208 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/msgpack.bundle                                                                    |  Bin 123096 -> 122584 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/packer.o                                                                          |  Bin 59880 -> 61728 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/packer_class.o                                                                    |  Bin 86712 -> 95664 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/packer_ext_registry.o                                                             |  Bin 7488 -> 8088 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/rbinit.o                                                                          |  Bin 3576 -> 2816 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/rmem.o                                                                            |  Bin 6336 -> 5224 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/unpacker.o                                                                        |  Bin 73576 -> 73648 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/unpacker_class.o                                                                  |  Bin 48168 -> 49912 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/ext/msgpack/unpacker_ext_registry.o                                                           |  Bin 9648 -> 9560 bytes
 {2.6.0 => 3.1.0}/gems/msgpack-1.6.0/lib/msgpack/msgpack.bundle                                                                    |  Bin 123096 -> 122584 bytes
 2.6.0/gems/racc-1.6.0/COPYING => /dev/null                                                                                        |   22 --
 2.6.0/gems/racc-1.6.0/ChangeLog => /dev/null                                                                                      |  846 ---------------------------------------------------------------------------
 2.6.0/gems/racc-1.6.0/README.ja.rdoc => /dev/null                                                                                 |   94 ---------
 2.6.0/gems/racc-1.6.0/README.rdoc => /dev/null                                                                                    |   83 --------
 2.6.0/gems/racc-1.6.0/TODO => /dev/null                                                                                           |    5 -
 2.6.0/gems/racc-1.6.0/bin/racc => /dev/null                                                                                       |  320 -----------------------------
 2.6.0/gems/racc-1.6.0/doc/en/NEWS.en.rdoc => /dev/null                                                                            |  282 -------------------------
 2.6.0/gems/racc-1.6.0/doc/en/grammar.en.rdoc => /dev/null                                                                         |  226 --------------------
 2.6.0/gems/racc-1.6.0/doc/en/grammar2.en.rdoc => /dev/null                                                                        |  219 --------------------
 2.6.0/gems/racc-1.6.0/doc/ja/NEWS.ja.rdoc => /dev/null                                                                            |  307 ---------------------------
 2.6.0/gems/racc-1.6.0/doc/ja/command.ja.html => /dev/null                                                                         |   94 ---------
 2.6.0/gems/racc-1.6.0/doc/ja/debug.ja.rdoc => /dev/null                                                                           |   36 ----
 2.6.0/gems/racc-1.6.0/doc/ja/grammar.ja.rdoc => /dev/null                                                                         |  348 -------------------------------
 2.6.0/gems/racc-1.6.0/doc/ja/index.ja.html => /dev/null                                                                           |   10 -
 2.6.0/gems/racc-1.6.0/doc/ja/parser.ja.rdoc => /dev/null                                                                          |  125 -----------
 2.6.0/gems/racc-1.6.0/doc/ja/usage.ja.html => /dev/null                                                                           |  414 -------------------------------------
 2.6.0/gems/racc-1.6.0/ext/racc/MANIFEST => /dev/null                                                                              |    4 -
 2.6.0/gems/racc-1.6.0/ext/racc/cparse/.sitearchdir.-.racc.time => /dev/null                                                       |    0
 2.6.0/gems/racc-1.6.0/ext/racc/cparse/Makefile => /dev/null                                                                       |  267 ------------------------
 2.6.0/gems/racc-1.6.0/ext/racc/cparse/cparse.bundle => /dev/null                                                                  |  Bin 38136 -> 0 bytes
 2.6.0/gems/racc-1.6.0/ext/racc/cparse/cparse.c => /dev/null                                                                       |  863 ----------------------------------------------------------------------------
 2.6.0/gems/racc-1.6.0/ext/racc/cparse/cparse.o => /dev/null                                                                       |  Bin 47624 -> 0 bytes
 2.6.0/gems/racc-1.6.0/ext/racc/cparse/extconf.rb => /dev/null                                                                     |    9 -
 2.6.0/gems/racc-1.6.0/lib/racc/compat.rb => /dev/null                                                                             |   33 ---
 2.6.0/gems/racc-1.6.0/lib/racc/cparse.bundle => /dev/null                                                                         |  Bin 38136 -> 0 bytes
 2.6.0/gems/racc-1.6.0/lib/racc/debugflags.rb => /dev/null                                                                         |   60 ------
 2.6.0/gems/racc-1.6.0/lib/racc/exception.rb => /dev/null                                                                          |   16 --
 2.6.0/gems/racc-1.6.0/lib/racc/grammar.rb => /dev/null                                                                            | 1118 ---------------------------------------------------------------------------------------------------
 2.6.0/gems/racc-1.6.0/lib/racc/grammarfileparser.rb => /dev/null                                                                  |  561 --------------------------------------------------
 2.6.0/gems/racc-1.6.0/lib/racc/info.rb => /dev/null                                                                               |   17 --
 2.6.0/gems/racc-1.6.0/lib/racc/iset.rb => /dev/null                                                                               |   92 ---------
 2.6.0/gems/racc-1.6.0/lib/racc/logfilegenerator.rb => /dev/null                                                                   |  212 -------------------
 2.6.0/gems/racc-1.6.0/lib/racc/parser-text.rb => /dev/null                                                                        |  637 --------------------------------------------------------
 2.6.0/gems/racc-1.6.0/lib/racc/parser.rb => /dev/null                                                                             |  632 --------------------------------------------------------
 2.6.0/gems/racc-1.6.0/lib/racc/parserfilegenerator.rb => /dev/null                                                                |  468 ------------------------------------------
 2.6.0/gems/racc-1.6.0/lib/racc/sourcetext.rb => /dev/null                                                                         |   35 ----
 2.6.0/gems/racc-1.6.0/lib/racc/state.rb => /dev/null                                                                              |  972 --------------------------------------------------------------------------------------
 2.6.0/gems/racc-1.6.0/lib/racc/statetransitiontable.rb => /dev/null                                                               |  311 ----------------------------
 2.6.0/gems/racc-1.6.0/lib/racc/static.rb => /dev/null                                                                             |    5 -
 2.6.0/gems/racc-1.6.0/lib/racc.rb => /dev/null                                                                                    |    6 -
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/Csio.o                                                                                |  Bin 7224 -> 5984 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/Makefile                                                                              |   75 +++----
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/amalloc.o                                                                             |  Bin 7312 -> 6880 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/basename.o                                                                            |  Bin 4128 -> 3368 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/css.o                                                                                 |  Bin 8696 -> 7160 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/docheader.o                                                                           |  Bin 6152 -> 4632 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/dumptree.o                                                                            |  Bin 11912 -> 10488 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/emmatch.o                                                                             |  Bin 9704 -> 9256 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/flags.o                                                                               |  Bin 8664 -> 7800 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/generate.o                                                                            |  Bin 101736 -> 124656 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/github_flavoured.o                                                                    |  Bin 10656 -> 8904 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/html5.o                                                                               |  Bin 3080 -> 2368 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/markdown.o                                                                            |  Bin 62680 -> 73104 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/mkdio.o                                                                               |  Bin 21832 -> 22448 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/mktags.o                                                                              |  Bin 6560 -> 13656 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/pgm_options.o                                                                         |  Bin 8456 -> 10712 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/rdiscount.bundle                                                                      |  Bin 106048 -> 120312 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/rdiscount.o                                                                           |  Bin 13248 -> 19408 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/resource.o                                                                            |  Bin 9736 -> 9048 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/setup.o                                                                               |  Bin 3056 -> 2296 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/tags.o                                                                                |  Bin 6720 -> 6456 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/toc.o                                                                                 |  Bin 10312 -> 8848 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/version.o                                                                             |  Bin 1832 -> 1152 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/xml.o                                                                                 |  Bin 6480 -> 5800 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/ext/xmlpage.o                                                                             |  Bin 7912 -> 6336 bytes
 {2.6.0 => 3.1.0}/gems/rdiscount-2.2.0.2/lib/rdiscount.bundle                                                                      |  Bin 106048 -> 120312 bytes
 {2.6.0/gems/sorbet-static-0.5.10461-universal-darwin-21 => 3.1.0/gems/sorbet-static-0.5.10461-universal-darwin-18}/libexec/sorbet |  Bin 16714048 -> 16714048 bytes
 {2.6.0 => 3.1.0}/gems/unf_ext-0.0.8.2/ext/unf_ext/Makefile                                                                        |   73 +++----
 {2.6.0 => 3.1.0}/gems/unf_ext-0.0.8.2/ext/unf_ext/unf.o                                                                           |  Bin 558416 -> 473864 bytes
 {2.6.0 => 3.1.0}/gems/unf_ext-0.0.8.2/ext/unf_ext/unf_ext.bundle                                                                  |  Bin 435704 -> 476720 bytes
 {2.6.0 => 3.1.0}/gems/unf_ext-0.0.8.2/lib/unf_ext.bundle                                                                          |  Bin 435704 -> 476720 bytes
 178 files changed, 372 insertions(+), 13619 deletions(-)
```
</details>

> require people on weird Ruby versions to manually run brew vendor-gems before brew will work as expected.

You can't, that's the point. Idea of the changes here is to allow a supported Ruby install gems for another Ruby. I could perhaps make it more generic rather than a fixed 3.1. Though 2.7 and 3.0 is totally untested while 3.1 has been tested (and can trivially be in CI if we want to keep it that way).

The idea I thought was to eventually allow > 2.6 to be used for at least some developer commands, which will need some form of vendored gems. It maybe doesn't need to be committed now if we're not ready, but I think it will at some point.

I don't think `brew vendor-gems` is something we want to encourage anyone to run themselves, so if you want people to jump through several hoops to run newer Ruby I don't think this is the way. `brew vendor-gems` isn't designed for users to use. It's mostly used in CI for dependabot updates and the rare occasion dependencies are added - it intentionally modifies files ready for committing. If you run Portable Ruby, `brew vendor-gems` will modify your `setup.rb` file to point to a different platform. Same if you run Ruby on Linux.

> I also don't like the idea of unconditionally forcing Ruby 3.1 to be installed outside of CI for this case.

When's the last time you ran `brew vendor-gems` locally? We're probably going to unconditionally require 3.1 for `brew typecheck` (Sorbet has already stopped support) and `brew style` (RuboCop support ends in a few months) eventually which is bigger issues. What direction are we heading here? I got a very different picture from the recent Slack discussion, so I think we should clarify this.

Reply:
> It is however not safe to symlink the entire directory:

Yeh, as you say, this all looks like stuff we shouldn't be committing and perhaps should just instead provide a way of e.g. symlinking the committed stuff and stripping out the object files as needed.

> The idea I thought was to eventually allow > 2.6 to be used for at least some developer commands, which will need some form of vendored gems.

Vendored as in committed into the repository? I'm not sure I see the logic here.

> When's the last time you ran brew vendor-gems locally?

Every month or so.

> We're probably going to unconditionally require 3.1 for brew typecheck (Sorbet has already stopped support) and brew style (RuboCop support ends in a few months) eventually which is bigger issues.

Seems good to me.

> What direction are we heading here? 

I don't see a different direction, same we've always had:
- we vendor the stuff that end-users need so they never have to run bundler
- we don't vendor anything only needed by maintainers/developer commands

As-per this: we should continue to vendor for portable-ruby and not for anything newer. If this requires some hacks to e.g. make `ruby.sh` know to run Bundler before attempting to run `brew.rb`: let's do that.

Reply:
Ok finally got back to looking at this after the drama of last week.

So something I guess to decide is actually how we are dealing with 2.6 gems vs 3.1 gems. This could mean:

- Same lockfile for both (allows us to symlink, but also means everything is pinned to support 2.6 which somewhat defeats the point)
- Separate lockfiles for both (continue 2.6 updates while allowing 3.1, but does mean double the amount of Dependabot updates - one PR for 2.6, another for 3.1)
- Freeze the 2.6 vendor directory as is, and make the lockfile target 3.1 only (and thus `brew install-bundler-gems` will also require Ruby 3.1)

Reply:
> also means everything is pinned to support 2.6 which somewhat defeats the point

agreed

> Freeze the 2.6 vendor directory as is, and make the lockfile target 3.1 only

don't really like how manual this would make security updates for 2.6

> Separate lockfiles for both (continue 2.6 updates while allowing 3.1, but does mean double the amount of Dependabot updates - one PR for 2.6, another for 3.1)

Seems like the best option?

I wonder if this would be worth splitting the Gemfiles themselves, too, so that we have a Gemfile for everything vendored and one for everything that's e.g. installed for developer commands?

Reply:
> I wonder if this would be worth splitting the Gemfiles themselves, too, so that we have a Gemfile for everything vendored and one for everything that's e.g. installed for developer commands?

Possibly. The tricky ones are when vendored dependencies are tied to non-vendored dependencies. e.g. sorbet-runtime <-> sorbet-static/tapioca, and the various RuboCop things that are vendored for some reason. So ultimately may mean drift in vendored dependencies anyway.

> we should continue to vendor for portable-ruby and not for anything newer

Given this, we could also introduce more groups like we did for Sorbet, so we're not installing `ronn` or `byebug` for someone who wants to just do a `brew install` on Ruby 3.1 (or even `brew style` to be honest).

I suppose it's a bit of a balance though, `brew install-bundler-gems` could be something that effectively covers the vendored parts by default and extends based on what you are using, but does mean it's less of a one-stop command (though maybe an `--all` would help with that).

Reply:
> So ultimately may mean drift in vendored dependencies anyway.

Yup, that makes sense.

> Given this, we could also introduce more groups like we did for Sorbet, so we're not installing ronn or byebug for someone who wants to just do a brew install on Ruby 3.1 (or even brew style to be honest).

As I've said before: I don't think the Sorbet groups work well currently at all so I'd be loathe to add more. Additionally, installing these gems (that are basically never updated) once is very little overhead. If anything I'd like to consider adding e.g. `stackprof` in here.

> I suppose it's a bit of a balance though, brew install-bundler-gems could be something that effectively covers the vendored parts by default and extends based on what you are using, but does mean it's less of a one-stop command (though maybe an --all would help with that).

Yeh, I think it's reasonable to decide what we want this to be for and adjust accordingly. Currently I see a few things:
- precache gem installation in e.g. CI
- part of vendoring gems
- ???

Reply:
> Perhaps we could move away from vendoring this?

This will mean removing any inline type signatures or `T.let` usage. So no, we can't.

>  If anything I'd like to consider adding e.g. stackprof in here.

I do agree with this. I don't think `Homebrew.install_gem!` should really be used beyond prototyping.